analyzers/src/SonarAnalyzer.CSharp/Rules/PrivateStaticMethodUsedOnlyByNestedClass.cs

```diff
@@ -0,0 +1,190 @@
+        SyntaxKind.ClassDeclaration,
+        SyntaxKind.StructDeclaration,
+        SyntaxKind.InterfaceDeclaration,
+        SyntaxKindEx.RecordClassDeclaration,
+        SyntaxKindEx.RecordStructDeclaration
+    };
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+            {
+                var declaredType = (TypeDeclarationSyntax)c.Node;
+
+                if (!IsPartial(declaredType)
+                    && HasNestedTypeDeclarations(declaredType)
+                    && PrivateStaticMethodsOf(declaredType) is { Length: > 0 } candidates)
+                {
+                    var methodReferences = TypesWhichUseTheMethods(candidates, declaredType, c.SemanticModel);
+
+                    foreach (var reference in methodReferences)
+                    {
+                        var typeToMoveInto = LowestCommonAncestorOrSelf(reference.Types);
+                        if (typeToMoveInto != declaredType)
+                        {
+                            var nestedTypeName = typeToMoveInto.Identifier.ValueText;
+                            c.ReportIssue(Diagnostic.Create(Rule, reference.Method.Identifier.GetLocation(), nestedTypeName));
+                        }
+                    }
+                }
+            },
+            AnalyzedSyntaxKinds);
+
+    private static MethodDeclarationSyntax[] PrivateStaticMethodsOf(TypeDeclarationSyntax type) =>
+        type.Members
+                .OfType<MethodDeclarationSyntax>()
+                .Where(x => IsPrivateAndStatic(x, type))
+                .ToArray();
+
+    private static bool HasNestedTypeDeclarations(TypeDeclarationSyntax type) =>
+        type.Members
+                .OfType<TypeDeclarationSyntax>()
+                .Any();
+
+    private static bool IsPrivateAndStatic(MethodDeclarationSyntax method, TypeDeclarationSyntax containingType) =>
+        method.Modifiers.Any(x => x.IsKind(SyntaxKind.StaticKeyword))
+        && ((HasAnyModifier(method, SyntaxKind.PrivateKeyword) && !HasAnyModifier(method, SyntaxKind.ProtectedKeyword))
+           || (!HasAnyModifier(method, SyntaxKind.PublicKeyword, SyntaxKind.ProtectedKeyword, SyntaxKind.InternalKeyword) && IsClassOrRecordClassOrInterfaceDeclaration(containingType)));
+
+    private static bool IsPartial(TypeDeclarationSyntax type) =>
+        type.Modifiers.Any(x => x.IsKind(SyntaxKind.PartialKeyword));
+
+    private static bool IsClassOrRecordClassOrInterfaceDeclaration(TypeDeclarationSyntax type) =>
+        type is ClassDeclarationSyntax or InterfaceDeclarationSyntax
+        || (RecordDeclarationSyntaxWrapper.IsInstance(type) && !((RecordDeclarationSyntaxWrapper)type).ClassOrStructKeyword.IsKind(SyntaxKind.StructKeyword));
```

Comment:
education: why not record structs?

Reply:
This method is only used when checking whether a method is private or not. The default accessibility for record class members is private, but for record structs (like all structs) it's internal. That's why it's treated separately here.